src/scripts/tbtcv2-rewards/rewards.ts

```diff
@@ -359,103 +340,99 @@ export async function calculateRewards() {
     requirements.set(IS_VERSION_SATISFIED, true)
-    if (upgradeCutoffDate < startRewardsTimestamp) {
-      // E.g. Feb interval
-      // ---older eligible tags---|----------latest tag only--------------|
-      // ---|---------------------|---------|-----------------------------|--->
-      // latest tag             cutoff     Feb1                         Feb28
 
-      // All the instances must run on the latest version during the rewards
-      // interval in Feb.
+    // if there is only one valid version, all the instances must run on it
+    if (validVersions.length === 1) {
       for (let i = 0; i < instances.length; i++) {
-        // May 1st 2024 distribution special case:
         if (
-          instances[i].buildVersion != "v2.0.0" ||
-          instances[i].buildVersion != "v2.0.1"
+          !checkMinorVersion(
+            validVersions[0].version,
+            instances[i].buildVersion
+          )
         ) {
           requirements.set(IS_VERSION_SATISFIED, false)
         }
       }
-    } else if (upgradeCutoffDate < endRewardsTimestamp) {
+    } else if (validVersions.length > 1) {
+      // for simplicity purposes, if there is several deadlines, upgradeCutoffDate is the
+      // least restrictive deadline.
+      const upgradeCutoffDate = validVersions[1].deadline
+      const latestClientTag = validVersions[0].version
+
+      // if we have more than one valid version, we need to check if the instances did the update
+      // before the cutoff date
+      //
       // E.g. Feb interval
       // -----older eligible tags-----|----latest tag only----|
       // ----|---------|--------------|-----------------------|--->
       // latest tag   Feb1          cutoff                   Feb28
-
-      // All the instances between (upgradeCutoffDate : endRewardsTimestamp]
-      // must run on the latest version
-      for (let i = instances.length - 1; i >= 0; i--) {
-        if (
-          instances[i].lastRegisteredTimestamp > upgradeCutoffDate &&
-          // May 1st 2024 distribution special case:
-          !(
-            instances[i].buildVersion === "v2.0.0" ||
-            instances[i].buildVersion === "v2.0.1"
-          )
-        ) {
-          // After the cutoff day a node operator still run an instance with an
-          // older version. No rewards.
-          requirements.set(IS_VERSION_SATISFIED, false)
-          // No need to check further since at least one instance run on the
-          // older version after the cutoff day.
-          break
-        } else {
-          // It might happen that a node operator stopped an instance before the
-          // upgrade cutoff date that happens to be right before the interval
-          // end date. However, it might still be eligible for rewards because
-          // of the uptime requirement.
-          // May 1st 2024 distribution special case:
+      if (upgradeCutoffDate < endRewardsTimestamp) {
```

Comment:
Just checking my understanding:

If `upgradeCutoffDate` is `< endRewardsTimestamp` but also `< startRewardsTimestamp` we are good because we make sure the instance runs `latestClientTag` after `upgradeCutoffDate` regardless of how far in the past `upgradeCutoffDate` actually is. Regarding this case, we can append new versions to the `tbtcValidVersions` constant (`gen_rewards_dist.js` file) and we don't need to remember to remove older versions from there.

However, if `upgradeCutoffDate >= endRewardsTimestamp` (next case) we are checking that an instance runs one of the `validVersionsNames` without further examining deadlines for each version. That means an instance can run a very old version that is still on the `tbtcValidVersions` list and still fulfill the version requirement. To avoid this case, we actually need to remember to update `tbtcValidVersions` on every script run.

Is that correct?

Reply:
> If upgradeCutoffDate is < endRewardsTimestamp but also < startRewardsTimestamp

Actually, this shouldn't happen in this new strategy that I'm proposing:

The constant `tbtcValidVersion` in `gen_rewards_dist.js` is intended to be modified on each rewards calculation. And it should include *only* the versions considered as valid for the distribution we are calculation at that moment. 

This case (`upgradeCutoffDate is < endRewardsTimestamp but also < startRewardsTimestamp`) is the same thing as saying that we have only one valid version, right?

For example, if we had a `v1.0` valid until Jan 25 (`cutoffDate` Jan25), and the next version is `v2.0`, the only valid version for February distribution should be `v2.0`, right?

```
|--v1.0 and v2.0 valid---|--only v2.0 valid----|
------------------------|---------|-----------|--->
                                    cutoff       Feb1          Feb28
```

So, the value for `tbtcValidVersions` will be `v2.0.0` for the Feb distribution. So the execution goes by the if statement on line 360.

Reply:
About the second case you are commenting: `upgradeCutoffDate >= endRewardsTimestamp`:

We are not including old versions on the `tbtcValidVersions`, only the valid ones for this distribution calculation.

An example could be:

For February  distribution, there are two valid versions: `v1.0.0` and version `v2.0.0`. The cutoff date of v1.0 is Mar 5th.

When calculating the rewards for February distribtuion, the variable `tbtcValidVersions` will be: `v2.0.0|v1.0.0_1709596800`. (The timestamp corresponds to Mar 5th.)

Reply:
Ok, thanks for the clarification. My whole question was actually: do we need to modify `tbtcValidVersions` on each rewards calculation? Now I see the answer is **yes**